Plot neural responses

fkang-pu opened the pull request:
**Summary:**

- Created media directory for storing atlas file. Used "mock.pickle" for now.

- Added a form (NeuronInputParamForm) in form.py with validation code to handle user input. Used hidden fields for both stim_neu_id and resp_neu_ids. 
 
- Added code in view.py to get validated values for form fields, and then call wormfunconn functions to generate data for plotting neural responses.
 
- New template file (plot_neural_responses.html) for form input and output.
 
- Minor changes to navbar.html

**Notes:**

- Ran Superlinter locally and passed. No new test code, although user input parameters can be easily verified as they are listed with graph in output section.

- Tried to get main steps of the web app. in place. Used "vanillaSelectBox" and form hidden fields for selecting neurons. I am open to suggestions or a different approach.
 
- Keep list_input_params.html and related code for now (as references for testing purpose). Plan to delete them later.
 
- Need suggestions on how to handle exceptions while calling wormfunconn functions. I used "raised ValueError..." in views.py, but not sure if it's the right way to handle it. The current version of wormfunconn only worked for **stim_type=rectangular.**

lparsons reviewed (changes requested):
This looks like a solid step forward, nice job Fan!

My one concern is the lack of tests, at least [tests that views return valid pages (200)](https://docs.djangoproject.com/en/3.2/intro/tutorial05/#test-a-view) and a simple request returns a plot (200).

As for the exception handling, I don't see there being much use to catching and re-raising Exceptions. With DEBUG on, we should see the exceptions, and with DEBUG off, they should be hidden. At some point, it would probably be good to report the errors on production in some way. Django docs on [error reporting](https://docs.djangoproject.com/en/3.2/howto/error-reporting/).

fkang-pu commented:
I am adding some test code, and will push new commits next Monday. If you haven't reviewed this PR, it's better to wait till next Monday.

fkang-pu commented:
Just pushed a commit. Added error catch for generating html output ((plot_div) from **ployly.offline.plot** function in views.py. Since **plot** function runs validation and including plot.js in output as default, the html output should render a graph is no error is raised.  Also modified and added code to test.py - check form status and expected output or errors.

Ran Super-linter locally and passed. However, GitHub Superlinter failed with the error: Docker pull failed with exit code 1.

lparsons commented:
@fkang and @jcmatese I believe the linting failure was just a temporary issue, however, I noticed that we were still on v3 of superlinter and decided to take the opportunity to bump to v4 and use the "slim" image which might help speed things up. If you use that when running locally, I've noticed it does speed things up a bit. I'm taking a look at the PR now.

lparsons reviewed (approved):
The tests look good to me, nice work Fan. I believe that linter error was just temporary. I did submit a PR to this branch to update to v4 and use the slim version of Super Linter. Optional if you want to include that or not.

lparsons commented:
@fkang and @jcmatese : While I'm always happy to have feedback from @hepcat72, I don't want to overburden people with reviews. Since Rob is already working on https://github.com/PrincetonUniversity/neuron-graph, we can let him off the hook for reviewing these PRs (though your welcome to if you feel you have the time Rob).

fkang-pu commented:
It's totally fine with me. I'll let Rob to decide if he wants to review this PR.  I can wait till tomorrow morning.

hepcat72 commented:
I'm fine reviewing. Just wrapped up the updates to the TB data submission PR.  If in the future, I'm slow to review, I wouldn't be offended if you merge without my review.  Or wait, or poke-tag me.  Whatever you want to do is fine by me.

hepcat72 commented on neuronsimulator/forms.py:
```diff
@@ -1,8 +1,97 @@
+            resp_neu_arr = cleaned_data.get("resp_neu_ids").split(",")
+            if stim_neu_id in resp_neu_arr:
+                raise forms.ValidationError(
+                    "Response neuron(s) included stimulated neuron, please re-select!"
```
NB.
```suggestion
                    "Response neuron(s) cannot include stimulated neuron(s), please make a different selection."
```

hepcat72 commented on neuronsimulator/forms.py:
```diff
@@ -1,8 +1,97 @@
+        dur = float(cleaned_data.get("dur"))
+        if t_max < dur:
+            raise forms.ValidationError(
+                "t_max value need to be larger than dur, please re-select!"
```
NB.
```suggestion
                "t_max must be larger than dur."
```

hepcat72 commented on neuronsimulator/forms.py:
```diff
@@ -1,8 +1,97 @@
+        """
+        if resp_neu_ids is None:
+            raise forms.ValidationError(
+                "need to select at least one response neuron (for testing!)"
```
NB.
```suggestion
                "You must select at least one response neuron (for testing)."
```

hepcat72 commented on neuronsimulator/forms.py:
```diff
@@ -1,8 +1,97 @@
+            )
+        else:
+            resp_neu_arr = resp_neu_ids.split(",")
+            for resp_neuron in resp_neu_arr:
+                if resp_neuron not in neurons:
+                    raise forms.ValidationError(
+                        "found invalid neuron in response neuron list, please re-select!"
+                    )
```
NB. Given multiple possibilities, I recommend collection all the errors first and report which ones are invalid.

```
            inv_neurs = []
            for resp_neuron in resp_neu_arr:
                if resp_neuron not in neurons:
                    inv_neurs.append(resp_neuron)
            if len(inv_neurs) > 0:
                raise forms.ValidationError(
                    f"Invalid response neuron(s) encountered: [{', '.join(inv_neurs)}]"
                )
```

hepcat72 commented on neuronsimulator/templates/home.html:
```diff
@@ -4,12 +4,20 @@
     <h3>Neuron Simulator Dashboard</h3>
     <br>
-    <h5>Note: placeholder; will add contents later</h5>
+    <h4>Overview</h4>
+    <p>Note: need feedback from the lab about this section </p>
```
NB. I try and avoid placeholders like this, but if you prefer to use them, I suggest employing one or more these tricks I have used in the past to easily find them in the future:
- Comment out the code - or put separate comments above and below.
- Use a keyword that you can use to find placeholders later, e.g. `#PLACEHOLDER#` or `#TODO#`.
- Create a github issue that either/both links to the line of code or add a link to the issue in the comment in the code.

hepcat72 commented on neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:
```diff
@@ -0,0 +1,140 @@
+            <br>
+            <div class="select-zone">
+                <label>Select a stimulated neuron:</label>
+                <select id="one-neuron">
+                    {% for neuron in neurons %}
+                        <option value="{{ neuron.name }}">{{ neuron.name }}</option>
+                    {% endfor %}
+                </select>
+            </div>
+
+            <div class="select-zone">
+                <label>Select one or more neuron(s) for calculating responses:</label>
+                <select id="multi-neurons" multiple="" size="12" style="display: inline-block;">
+                    {% for neuron in neurons %}
+                        <option value="{{ neuron.name }}">{{ neuron.name }}</option>
+                    {% endfor %}
+                </select>
```
Q. Why are these select lists not located inside `form` tags? Are the form tags in a parent template or are these used in javascript?

hepcat72 commented on neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:
```diff
@@ -0,0 +1,140 @@
+                            {% endif %}
+                            {{ field }}
+                            {% if field.help_text %}
+                                <small class="form-text">{{ field.help_text }}</small>
```
C. I often use `<small>` tags like this, and have gotten feedback that it should be done with CSS.

hepcat72 commented on neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:
```diff
@@ -0,0 +1,140 @@
+
+            let options = document.getElementById("multi-neurons").selectedOptions;
+            let multiNeuronsChoices = Array.from(options).map(({ value }) => value);
+            //console.log(multi-neuronsChoices);
```
NB. Remove commented debug print.

hepcat72 commented on neuronsimulator/views.py:
```diff
@@ -1,12 +1,140 @@
+        stim_type = params["stim_type"]
+        dur = float(params["dur"])
+
+        # print("stim_neu_id", stim_neu_id)
+        # print("resp_neu_ids", resp_neu_ids)
```
NB. Remove commented debug prints.

hepcat72 commented on neuronsimulator/views.py:
```diff
@@ -1,12 +1,140 @@
+            raise FileNotFoundError("Input Atlas file was not found")
+
+        # calculate dt from t_max and nt
+        dt = t_max / nt
```
NB. I know that the form has a min value of 1 in the form validation, but for security concerns (because client-side validation can be circumvented), I think doing a server-side code check that `nt` is greater than 0 would prevent an exception upon invalid input.

hepcat72 commented on neuronsimulator/views.py:
```diff
@@ -1,12 +1,140 @@
+        except Exception as e:
+            raise ValueError("Got error with get_standard_stimulus:", e)
+
+        # print("The size of stim ndarray is:", stim.size)
```
NB. Remove commented debug print.

hepcat72 replied:
here and elsewhere (e.g. line 76)

hepcat72 replied:
... or wrap them in a conditional on the value of `settings.DEBUG`.

hepcat72 reviewed (approved):
This is great work Fan.  Feel free to ignore all my knits.  Approved.

fkang-pu commented:
@hepcat72 Rob, thanks a lot for you detailed review and suggestions, especially the code for invalid reps_neurons.  I will make changes accordingly. Just address your main comments here instead of one-by-one: 

- **form filed names and descriptions**: I used those listed in Francesco's` wormfunconn` package. I won't make changes until I get feedback from him. His code can only handle one stim_type right now, so I expect more changes will be made on his end.

- **debug code**: plan to do a final clean up when we are close to move the code to production. In the next stage of development/testing, I'll still need those lines of debug code.

- **stim_neu_id and resp_neu_ids selections**: I want to use JavaScript code from VanillaSelectBox (a nice web front for multiple selections with auto suggestion feature). I tried different options and also had a discussion with John. The approach I used is the best I can come up with right now.

fkang-pu commented:
@lparsons I made some minor changes based on Rob's suggestions. Tested again with test.py and from web front. I had to fetch/merge the changes from the remote branch (from your PR7 initially) to my local branch before pushing new commits to GitHub. You may want to take a quick look before I merge this PR to main. Thanks.

lparsons commented on neuronsimulator/forms.py:
```diff
@@ -77,17 +75,20 @@ def clean_resp_neu_ids(self):
-                    raise forms.ValidationError(
-                        "found invalid neuron in response neuron list, please re-select!"
-                    )
+                    invalid_neurons.append(resp_neuron)
```
Just a couple of [options that are a bit more "pythonic"](https://www.geeksforgeeks.org/python-difference-two-lists/).

Use set difference operation:
```python
invalid_neurons = list( set(resp_neu_arr) - set(neurons) )
```

Using list comprehension:
```python
invalid_neurons = [i for i in resp_neu_arr if i not in neurons]
```

fkang-pu replied:
@lparsons Nice suggestions., thanks. I tested both in Django shell and opt to use 2nd one. Just pushed another commit.

lparsons reviewed (commented):
Looks great @fkang-pu. I had one suggestion for how you could simplify the code a bit using a more ["pythonic" style](https://www.udacity.com/blog/2020/09/what-is-pythonic-style.html).
